cpu/esp32/esp-bootloader-reset/reset.c

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (C) 2023 Benjamin Valentin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_esp32
+ * @{
+ *
+ * @file
+ * @brief       Trigger reset to the bootloader stored in the internal boot ROM
+ *              memory.
+ *
+ *              This will start the USB/UART bootloader.
+ *
+ * @author      Benjamin Valentin <benpicco@googlemail.com>
+ *
+ * @}
+ */
+
+#include "rom/usb/chip_usb_dw_wrapper.h"
+#include "rom/usb/usb_persist.h"
+#include "soc/rtc_cntl_reg.h"
+#include "periph/pm.h"
+
+void __attribute__((weak)) usb_board_reset_in_bootloader(void)
+{
+    chip_usb_set_persist_flags(USBDC_PERSIST_ENA);
```

Comment:
Wow, where did you find a hint about this functionality? I was always looking for a way to restart the ESP32x SoCs in download mode. I'm really impressed.

The question is how to do the same for ESP32-C3.

Reply:
I was searching for reboot and bootloader in `esp32_sdk` and found something that looked good in `esp_usb_console_before_restart()`.

AFAIK ESP32-C3 doesn't have a USB controller, only the debug serial/JTAG core, which already does a reset to bootloader via USB.

Reply:
> AFAIK ESP32-C3 doesn't have a USB controller, only the debug serial/JTAG core

That's clear, but I didn't know that it doesn't require something special and you can just use the standard flash method using. I thought the special reset sequence from the TRM is required? But obviously it is sufficient to connect the D+ and D-  and to say `make flash ...`. That's great. I didn't know that.

Reply:
It has a USB implementation in ROM:

```
[17517.347080] usb 1-1.3: USB disconnect, device number 34
[17517.830175] usb 1-1.3: new full-speed USB device number 35 using ehci-pci
[17517.940703] usb 1-1.3: New USB device found, idVendor=1209, idProduct=7d01, bcdDevice= 1.00
[17517.940714] usb 1-1.3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[17517.940717] usb 1-1.3: Product: esp32s2-wemos-mini
[17517.940720] usb 1-1.3: Manufacturer: RIOT-os.org
[17517.940722] usb 1-1.3: SerialNumber: A591F27212482C31
[17517.942123] cdc_acm 1-1.3:1.0: ttyACM0: USB ACM device
# do bootloader reset
[17524.259036] usb 1-1.3: USB disconnect, device number 35
[17524.486257] usb 1-1.3: new full-speed USB device number 36 using ehci-pci
[17524.605137] usb 1-1.3: New USB device found, idVendor=303a, idProduct=0002, bcdDevice= 7.23
[17524.605147] usb 1-1.3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[17524.605151] usb 1-1.3: Product: ESP32-S2
[17524.605154] usb 1-1.3: Manufacturer: Espressif
[17524.605156] usb 1-1.3: SerialNumber: 0
```

Reply:
Hm, ...., this is exactly what I supposed but when I tested it with my ESP32-S2 DevKit, there was no USB interface after the reset in download mode. Maybe I did something wrong.